Add helm and Github Actions CI for Kubernetes dev

hamishfagg opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Adds working implementation in Github Actions of build/test/release/deploy for:

- Backend with Envoy proxy
- Frontend
- Media server

Helm charts have been added in the `/deployment` folder for deploying the above services to the Kubernetes dev environment.
This replaced the old Gitlab CI script that was at `/app/.gitlab-ci.yml`

This PR also adjusts the naming of some docker/env files to avoid confusion. In general things are now named:

- "local" - for running services on your local machine for development
- "dev" - for deployment to the dev Kubernetes environment
- "prod" - for deployment to the prod (and eventually, staging) Kubernetes environments

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on the web frontend)
If you need help with any of these, please ask :)
--->
**Backend checklist**
- [x] All tests pass
- [x] Run the backend locally and it works

**Web frontend checklist**
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] All tests pass
- [x] Clicked around my changes running locally and it works


<!---
Remember to request review from couchers-org/web, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

hamishfagg committed:
Add helm charts and github actions CI for backend/proxy, frontent, and media

hamishfagg committed:
rename helm master files to dev. Update readmes

hamishfagg committed:
Fixed actions path and helm script file names

AndrewFarley reviewed (dismissed):
Reviewed, after some discussion and changes all LGTM.

aapeliv commented:
This seems to break the config used for local dev using docker compose?

IVData commented:
> This seems to break the config used for local dev using docker compose?

It's working for me following the readme in /app. What problem are you having?

aapeliv commented:
@IVData: I don't know where your comment went; but it's not working when I follow the readme in `/app`:

```
backend_1   | 2022-05-04 23:35:46,965: couchers.jobs.handlers: Sending email with subject '[DEV] You have 124 unseen messages in 3 chats on Couchers.org!' to '...@couchers.org'                                                                                                       
backend_1   | 2022-05-04 23:35:46,966: couchers.jobs.worker: encoding with 'idna' codec failed (UnicodeError: label empty or too long)                                                                                                                                                    
backend_1   | Traceback (most recent call last):                                                                                                                                                                                                                                          
backend_1   |   File "/usr/local/lib/python3.8/encodings/idna.py", line 165, in encode                                                                                                                                                                                                    
backend_1   |     raise UnicodeError("label empty or too long")                                                                                                                                                                                                                           
backend_1   | UnicodeError: label empty or too long                                                                                                                                                                                                                                       
backend_1   |                                                                                                                                                                                                                                                                             
backend_1   | The above exception was the direct cause of the following exception:                                                                                                                                                                                                        
backend_1   |                                                                                                                                                                                                                                                                             
backend_1   | Traceback (most recent call last):                                                                                                                                                                                                                                          
backend_1   |   File "/app/src/couchers/jobs/worker.py", line 55, in process_job                                                                                                                                                                                                          
backend_1   |     ret = func(message_type.FromString(job.payload))                                                                                                                                                                                                                        
backend_1   |   File "/app/src/couchers/jobs/handlers.py", line 52, in process_send_email                                                                                                                                                                                                 
backend_1   |     email = sender(                                                                                                                                                                                                                                                         
backend_1   |   File "/app/src/couchers/email/smtp.py", line 27, in send_smtp_email                                                                                                                                                                                                       
backend_1   |     with smtplib.SMTP(config["SMTP_HOST"], config["SMTP_PORT"]) as server:                                                                                                                                                                                                  
backend_1   |   File "/usr/local/lib/python3.8/smtplib.py", line 255, in __init__                                                                                                                                                                                                         
backend_1   |     (code, msg) = self.connect(host, port)                                                                                                                                                                                                                                  
backend_1   |   File "/usr/local/lib/python3.8/smtplib.py", line 339, in connect                                                                                                                                                                                                          
backend_1   |     self.sock = self._get_socket(host, port, self.timeout)                                                                                                                                                                                                                  
backend_1   |   File "/usr/local/lib/python3.8/smtplib.py", line 310, in _get_socket                                                                                                                                                                                                      
backend_1   |     return socket.create_connection((host, port), timeout,                                                                                                                                                                                                                  
backend_1   |   File "/usr/local/lib/python3.8/socket.py", line 787, in create_connection                                                                                                                                                                                                 
backend_1   |     for res in getaddrinfo(host, port, 0, SOCK_STREAM):                                                                                                                                                                                                                     
backend_1   |   File "/usr/local/lib/python3.8/socket.py", line 918, in getaddrinfo                                                                                                                                                                                                       
backend_1   |     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):                                                                                                                                                                                                 
backend_1   | UnicodeError: encoding with 'idna' codec failed (UnicodeError: label empty or too long)                                                                                                                                                                                     
backend_1   | 2022-05-04 23:35:46,967: couchers.jobs.worker: Job #13 error on try number 1, next try at 2022-05-04 23:36:16.165086+00:00                                                                                                                                                  
backend_1   | 2022-05-04 23:35:46,972: couchers.jobs.worker: Job #14 grabbed                                                                                                                                                                                                              
```

Also the proxy is returning `503 UF` and the web instance can't connect to the backend.

darrenvong commented on app/web/.env.development:
```diff
@@ -1,6 +1,6 @@
 NEXT_PUBLIC_COUCHERS_ENV=dev
-NEXT_PUBLIC_API_BASE_URL=http://localhost:8888
-NEXT_PUBLIC_MEDIA_BASE_URL="http://localhost:5000"
```
I think we should keep these to point to `localhost` so those of us who want to run everything locally don't have to update this separately, which seems suiting for dev environment variables

hamishfagg replied:
I agree, please see below.
The default when running `docker-compose up` will now be to use these localhost URLs

aapeliv replied:
(oops, edited your comment)

hamishfagg replied:
Was missing a file here to allow local dev to work (.env.local). Now tested working.

hamishfagg committed:
add local env files for local dev

hamishfagg commented:
I've moved some env files around and added a set of 'local' env files for local development, and left 'dev' for the remote development infrastructure to stay closer to the standard layout of things.

For me at least, the readme and docker-compose for local development work correctly _except for media requests_. I'm unsure how media requests worked previously for local dev, since the media server does not reply with any CORS headers and so the browser blocks the responses. Did this work previously?

hamishfagg commented:
Have just confirmed that when using docker-compose.yml in the develop branch, requests to the media server throw the same CORS errors. So as far as I can tell, this branch is in the same state as develop, as far as local dev goes.

darrenvong commented:
The media server was always broken locally for me... 😅  I remember having to add a CORS header hack temporarily to get it working

hamishfagg commented:
> The media server was always broken locally for me... 😅 I remember having to add a CORS header hack temporarily to get it working

Where did you add that, out of curiosity?

hamishfagg committed:
remove unused dockerfile for frontend and fix up GH actions workflows for env changes

hamishfagg commented:
Updated again to fix up all of the remaining issues related to renaming things.

Would be great if someone can test the local environment again to make sure it's not just working for me (@Aapeli ?)

aapeliv commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,81 @@
+          black --check --diff .
+
+
+  release:
+    runs-on: [self-hosted, master]
+    needs: [test-backend, test-backend-format]
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ./.github/actions/retag-release-ecr
+        with:
+          module-name: backend
```
Will this now include all the coverage reports, etc?

hamishfagg replied:
No, but since the backend is being replaced I'm not sure it worth standing that up now

aapeliv commented on .github/workflows/backend.yml:
```diff
@@ -0,0 +1,81 @@
+        run: |
+          curl -sLO https://github.com/Couchers-org/resources/raw/a664750dc60771ec5080ac4753d9a27e6ec544b9/timezone_areas/timezone_areas.sql.zst
+          zstd -d timezone_areas.sql.zst -o resources/timezone_areas.sql
+          docker-compose -f ../docker-compose.test.yml up -d postgres_tests
```
Is it possible for this not to start on time for the tests?

hamishfagg replied:
Possible perhaps? I've not had it happen yet>
Again because this is a temp solution to stand up the old backend.
We could just add a sleep? Installing requirements below takes a while before tests start

aapeliv commented on app/backend.dev.env:
```diff
@@ -18,26 +18,26 @@ STRIPE_RECURRING_PRODUCT_ID=price_...
-MEDIA_SERVER_SECRET_KEY=91e29bbacc74fa7e23c5d5f34cca5015cb896e338a620003de94a502a461f4bc
-MEDIA_SERVER_BEARER_TOKEN=c02d383897d3b82774ced09c9e17802164c37e7e105d8927553697bf4550e91e
-MEDIA_SERVER_BASE_URL=http://127.0.0.1:5000
+MEDIA_SERVER_SECRET_KEY=dea94c7666a54dab01cce89fd80f0320
```
Are these the real secrets to the live media dev server?

hamishfagg replied:
:)
I've already rotated them and pushed an update to remove these ones.

aapeliv commented on app/backend/requirements.txt:
```diff
@@ -1,5 +1,5 @@
 #
```
Accidentally downgrading stuff here?

hamishfagg replied:
have reverted these

aapeliv commented on app/backend/requirements.in:
```diff
@@ -3,7 +3,7 @@ autoflake
 boto3
 geoalchemy2
-grpcio
+grpcio<=1.43.2
```
Why?

hamishfagg replied:
have reverted

aapeliv commented on app/docker-compose.yml:
```diff
@@ -28,17 +23,25 @@ services:
       - 1751:1751
       - 1752:1752
       - 1753:1753
+      - 8888:8888
+      - 9901:9901
```
Not needed for backend

hamishfagg replied:
Backend and proxy are put on the same network in this file now. This is due to the setup in kubernetes, without having to edit envoy.yaml between kubernetes and local deployment.
Easiest option is to just place the proxy and backend containers on the same IP
So these ports are still for proxy, but must be defined here as we're using the backend service's network.

aapeliv commented on app/proxy/envoy.yaml:
```diff
@@ -114,7 +117,7 @@ static_resources:
       - lb_endpoints:
         - endpoint:
-            address: { socket_address: { address: backend, port_value: 1751 }}
+            address: { socket_address: { address: 127.0.0.1, port_value: 1751 }}
```
Frontend (after changing back to `localhost` URLs) doesn't connect to the backend; I think it's because of this line.

hamishfagg replied:
There is a bug in envoy that requires this for kubernetes.
docker-compose.yaml has been edited for this to work in local deployment too. I will double check this isnt broken

hamishfagg replied:
This does work when tested locally.
Backend may time out trying to connect while postgres starts up and is populated. You may need to run docker-compose a second time. Any help making backend wait for postgres is appreciated but its not a big problem for now to just run it again once postgres has been initialised.

aapeliv reviewed (commented):
I would personally prefer the deployment stuff to be put in the `/app` directory, but that's just a preference!

hamishfagg committed:
fix local frontend and dev deployment, revert requirements changes

hamishfagg committed:
remove secrets (theyve already been rotated, dw)

darrenvong commented on app/web/Dockerfile:
```diff
@@ -38,4 +38,4 @@ ENV PORT 3000
 ENV NEXT_TELEMETRY_DISABLED 1
 
-CMD ["node_modules/.bin/next", "start"]
+CMD ["yarn", "build", "&&", "yarn", "install", "--production", "--ignore-scripts", "--prefer-offline", "&&", "node_modules/.bin/next", "start"]
```
Could we not skip `yarn build` here since it's already been run in a previous image layer?

hamishfagg replied:
We need to build at run-time because of how next handles env vars.
Our choices are either to have 3 separate containers (dev, staging, prod) and run `yarn build` at build time, or have one container and build at run-time.
The first build further up the file seems to speed up subsequent builds so it makes starting the container faster.

darrenvong replied:
Sounds like that first build must be caching something that helps re-running `yarn build` again.
Curious if previous build's cache would speed up the current build. If yes, would it be faster to then restore the cache rather than running `yarn build` twice, as that step in itself seems to take ~5-6 mins

darrenvong commented:
> > The media server was always broken locally for me... 😅 I remember having to add a CORS header hack temporarily to get it working
> 
> Where did you add that, out of curiosity?

I had to add this at the end of the route functions as a hack before the `return app` line:

```python
    @app.after_request
    def add_header(response: Response):
        response.headers['Access-Control-Allow-Origin'] = '*'
        return response
```

hamishfagg committed:
move to a single web image, and add deployment to staging env

hamishfagg commented:
(rebased onto develop)

hamishfagg committed:
fix image tags

hamishfagg committed:
add staging urls to proxy def

hamishfagg committed:
add resource limits to deployments

hamishfagg committed:
update urls to point to new staging

hamishfagg committed:
remove SameSite=strict from dev env sesh cookie

hamishfagg committed:
remove SameSite=strict from dev env sesh cookie

hamishfagg committed:
remove frontend files

AndrewFarley committed:
missing image in readme, minor readme update

AndrewFarley committed:
fixing cors issues on media
